src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -462,61 +462,52 @@ private static TypeSymbol MakeMergedTupleType(ArrayBuilder<DeconstructionVariabl
+            var namesBuilder = ArrayBuilder<string>.GetInstance(count);
             foreach (var variable in variables)
             {
                 if (variable.HasNestedVariables)
                 {
                     var nestedTuple = DeconstructionVariablesAsTuple(variable.Syntax, variable.NestedVariables, diagnostics, hasErrors);
                     valuesBuilder.Add(nestedTuple);
                     typesBuilder.Add(nestedTuple.Type);
-                    if (inferNames)
-                    {
-                        namesBuilder.Add(null);
-                    }
+                    namesBuilder.Add(null);
                 }
                 else
                 {
                     var single = variable.Single;
                     valuesBuilder.Add(single);
                     typesBuilder.Add(single.Type);
-                    if (inferNames)
-                    {
-                        namesBuilder.Add(ExtractDeconstructResultElementName(single));
-                    }
+                    namesBuilder.Add(ExtractDeconstructResultElementName(single));
                 }
                 locationsBuilder.Add(variable.Syntax.Location);
             }
 
-            ImmutableArray<string> tupleNames;
-            if (inferNames)
-            {
-                var uniqueFieldNames = PooledHashSet<string>.GetInstance();
-                RemoveDuplicateInferredTupleNames(namesBuilder, uniqueFieldNames);
-                uniqueFieldNames.Free();
-                tupleNames = namesBuilder.ToImmutableAndFree();
-            }
-            else
-            {
-                tupleNames = default(ImmutableArray<string>);
-            }
+            var uniqueFieldNames = PooledHashSet<string>.GetInstance();
+            RemoveDuplicateInferredTupleNames(namesBuilder, uniqueFieldNames);
+            uniqueFieldNames.Free();
+            ImmutableArray<string> tupleNames = namesBuilder.ToImmutableAndFree();
+
+            // only track the inferred positions in C# 7
+            var inferredPositions = this.Compilation.LanguageVersion.DisallowInferredTupleElementNames()
+                ? tupleNames.SelectAsArray(n => n != null)
+                : default(ImmutableArray<bool>);
 
             var type = TupleTypeSymbol.Create(syntax.Location,
                 typesBuilder.ToImmutableAndFree(), locationsBuilder.ToImmutableAndFree(),
                 tupleNames, this.Compilation,
-                shouldCheckConstraints: !hasErrors, syntax: syntax, diagnostics: hasErrors ? null : diagnostics);
+                shouldCheckConstraints: !hasErrors, inferredPositions: inferredPositions,
+                syntax: syntax, diagnostics: hasErrors ? null : diagnostics);
 
-            var inferredPositions = inferNames ? tupleNames.SelectAsArray(n => n != null) : default(ImmutableArray<bool>);
             return new BoundTupleLiteral(syntax, tupleNames, inferredPositions, arguments: valuesBuilder.ToImmutableAndFree(), type: type);
```

Comment:
This should probably save the positions even if it isn't in the type, however, this is probably not observable until we permit tuple element names on the left of a deconstruction.